google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java

```diff
@@ -0,0 +1,578 @@
+            // the current response, and we're expecting the rest of the chunked cells in the
+            // following ReadChangeStream response.
+            validate(
+                index == dataChange.getChunksCount() - 1,
+                "AWAITING_CELL_VALUE: Current SetCell is not chunked.");
+            return AWAITING_CELL_VALUE;
+          }
+          // Case 2: Current SetCell is not chunked.
+          builder.finishCell();
+          return checkAndFinishMutationIfNeeded(dataChange, index + 1);
+        }
+      };
+
+  /**
+   * A state that represents a completed change stream record. It prevents new change stream records
+   * from being read until the current one has been consumed.
+   *
+   * <dl>
+   *   <dt>Valid exit states:
+   *   <dd>{@link ChangeStreamStateMachine#AWAITING_NEW_STREAM_RECORD}.
+   * </dl>
+   */
+  private final State AWAITING_STREAM_RECORD_CONSUME =
+      new State() {
+        @Override
+        State handleHeartbeat(ReadChangeStreamResponse.Heartbeat heartbeat) {
+          throw new IllegalStateException(
+              "AWAITING_STREAM_RECORD_CONSUME: Skipping completed change stream record.");
+        }
+
+        @Override
+        State handleCloseStream(ReadChangeStreamResponse.CloseStream closeStream) {
+          throw new IllegalStateException(
+              "AWAITING_STREAM_RECORD_CONSUME: Skipping completed change stream record.");
+        }
+
+        @Override
+        State handleMod(ReadChangeStreamResponse.DataChange dataChange, int index) {
+          throw new IllegalStateException(
+              "AWAITING_STREAM_RECORD_CONSUME: Skipping completed change stream record.");
+        }
+      };
+
+  /**
+   * Check if we should continue handling mods in the current DataChange or wrap up. There are 3
+   * cases:
+   *
+   * <ul>
+   *   <li>1) index < dataChange.getChunksCount() -> continue to handle the next mod.
+   *   <li>2_1) index == dataChange.getChunksCount() && dataChange.done == true -> current change
+   *       stream mutation is complete. Wrap it up and return {@link
+   *       ChangeStreamStateMachine#AWAITING_STREAM_RECORD_CONSUME}.
+   *   <li>2_2) index == dataChange.getChunksCount() && dataChange.done != true -> current change
+   *       stream mutation isn't complete. Return {@link ChangeStreamStateMachine#AWAITING_NEW_MOD}
+   *       to wait for more mods in the next ReadChangeStreamResponse.
+   * </ul>
+   */
+  private State checkAndFinishMutationIfNeeded(
+      ReadChangeStreamResponse.DataChange dataChange, int index) {
+    validate(
+        0 <= index && index <= dataChange.getChunksCount(),
+        "checkAndFinishMutationIfNeeded: index out of bound.");
+    // Case 1): Handle the next mod.
+    if (index < dataChange.getChunksCount()) {
+      return AWAITING_NEW_MOD.handleMod(dataChange, index);
+    }
+    // We have finished handling all the mods in this DataChange.
```

Comment:
Should be 'in a logical mutation'

Reply:
I realized that this is the right comment but is confusing. Updated the comments.  Does it make more sense now?